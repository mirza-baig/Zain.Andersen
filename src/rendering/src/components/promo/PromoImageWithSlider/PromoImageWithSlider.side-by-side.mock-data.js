const sideBySideData = {
    "rendering": {
     "componentName": "PromoImageWithSlider",
     "dataSource": "/sitecore/content/AndersenCorporation/AndersenWindows/AndersenWindows/Home/Components/Promo Image With Slider/Data/Promo Image With Slider",
    },
     "fields": {
       "description": {
         "value": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
       },
       "imageCaption": {
         "value": "Image caption goes here."
       },
       "layoutStyle": {
         "id": "b68872e6-5040-4edb-b752-01908a65453d",
         "url": "http://localhost/sitecore/login/sitecore/system/Settings/Feature/EnterpriseWeb/Enums/Promo-Image-With-Slider-Style/Full-Width",
         "name": "Full Width",
         "displayName": "Full Width",
         "fields": {
           "Value": {
             "value": "side-by-side"
           }
         },
         "templateId": "d2923fee-da4e-49be-830c-e27764dfa269",
         "templateName": "Enum"
       },
       "leftImage": {
         "value": {
           "src": "https://source.unsplash.com/random/795x447?before",
           "alt": "andersen big doors looking from the inside out to patio garden area",
           "width": "392",
           "height": "392"
         }
       },
       "leftImageLabel": {
         "value": "Before"
       },
       "rightImage": {
         "value": {
           "src": "https://source.unsplash.com/random/795x447?after",
           "alt": "woman in entrance of home with andersen black framed door open",
           "width": "392",
           "height": "392"
         }
       },
       "rightImageLabel": {
         "value": "After"
       },
       "componentSpacing": null,
       "sectionId": {
         "value": ""
       },
       "headlineLevel": {
         "id": "0f556f3a-bbad-4aa5-952d-b79003b39cd6",
         "url": "http://localhost/sitecore/login/sitecore/system/Settings/Foundation/EnterpriseWeb/Enums/Headline-Levels/Heading-2",
         "name": "Heading 2",
         "displayName": "Heading 2",
         "fields": {
           "Value": {
             "value": "h2"
           }
         },
         "templateId": "d2923fee-da4e-49be-830c-e27764dfa269",
         "templateName": "Enum"
       },
       "headlineText": {
         "value": "Headline Text"
       },
       "eventName": {
         "value": ""
       },
       "eventType": {
         "value": ""
       },
       "eventZone": {
         "value": ""
       }
     }
 };

 export default sideBySideData;
