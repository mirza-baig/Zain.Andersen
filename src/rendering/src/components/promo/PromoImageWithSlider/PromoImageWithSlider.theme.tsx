import { ThemeFile } from 'lib/context/ThemeContext';

export const PromoImageWithSliderTheme: ThemeFile = {
  aw: {
    classes: {
      sideBySide: {
        headline: 'text-theme-text text-sm-m md:text-m  font-heavy',
        promoContent: 'col-span-12 ml:col-span-3 order-last ml:order-first flex flex-col',
        promoImgSlider: 'col-span-12 ml:col-span-9 ',
      },
      fullWidth: {
        headline: 'text-theme-text text-sm-m md:text-m  font-heavy',
        promoContent: 'col-span-12 order-last',
        promoImgSlider: 'col-span-12',
      },
      expandButton: {
        wrapper: 'flex flex-row md:flex-row-reverse items-center gap-x-xxs  group',
        icon: 'flex h-m w-m items-center justify-center rounded-full border-2 border-primary hover:bg-primary',
        text: 'text-text-link font-sans font-regular group-hover:underline underline-offset-4',
      },
      caption: 'md:mb-0 ',
      imageLabels:
        'top-[30px] z-10 inline-block py-s text-center text-sm-xxs md:text-xs font-sans font-heavy text-black md:absolute md:w-[160px] md:bg-black md:text-white uppercase md:hidden',
      rangewrapper: 'flex items-center justify-center bg-light-gray px-m',
      bodyCopy:
        'col-span-12 md:col-span-10 text-theme-body text-body mb-s font-regular mt-xxs md:mt-s',
    },
  },
  rba: {
    classes: {
      sideBySide: {
        headline: 'text-theme-text text-sm-m md:text-m font-medium ',
        promoContent:
          'col-span-12 ml:col-span-4 order-last ml:order-first flex flex-col justify-between',
        promoImgSlider: 'col-span-12 ml:col-span-8',
      },
      fullWidth: {
        headline: 'text-theme-text text-sm-m md:text-l font-medium ',
        promoContent: 'col-span-12 order-last',
        promoImgSlider: 'col-span-12',
      },
      expandButton: {
        wrapper: 'flex flex-row items-center gap-x-xxs mb-xxs md:mb-0 ',
        icon: '',
        text: 'text-small font-heavy',
      },
      caption: '',
      bodyCopy: 'col-span-12 md:col-span-10 text-theme-body text-body mb-s font-regular',
      rangewrapper: 'flex items-center justify-center bg-light-gray px-xxs',
      imageLabels:
        'top-[30px] z-10 inline-block py-s text-center text-sm-xxs md:text-xs font-serif font-bold text-black md:absolute md:w-[160px] md:bg-black md:text-white uppercase md:hidden',
    },
  },
};
