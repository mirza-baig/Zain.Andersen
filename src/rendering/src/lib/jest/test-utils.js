// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
