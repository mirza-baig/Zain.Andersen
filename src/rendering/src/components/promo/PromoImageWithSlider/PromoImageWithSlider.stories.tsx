// Global
import { Story, Meta } from '@storybook/react';
// Local
import PromoImageWithSlider, { PromoImageWithSliderProps } from './PromoImageWithSlider';
import defaultData from './PromoImageWithSlider.full-width.mock-data';
import sideBySideData from './PromoImageWithSlider.side-by-side.mock-data';

export default {
  title: 'Promo/PromoImageWithSlider',
  component: PromoImageWithSlider,
} as Meta;
const Template: Story<PromoImageWithSliderProps> = (props) => <PromoImageWithSlider {...props} />;
export const FullWidth = Template.bind({});
FullWidth.args = defaultData;

export const SideBySide = Template.bind({});
SideBySide.args = sideBySideData;
