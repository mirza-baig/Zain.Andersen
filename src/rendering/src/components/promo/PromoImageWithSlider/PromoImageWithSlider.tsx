// Global
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { useTheme } from 'lib/context/ThemeContext';
import { Text, withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
import { Component } from 'src/helpers/Component';
import { PromoImageWithSliderTheme } from './PromoImageWithSlider.theme';
import { KeyboardEvent, useEffect, useRef, useState } from 'react';
import ReactBeforeSliderComponent from 'react-before-after-slider-component';
import { getEnum } from 'lib/utils';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { Headline } from 'src/helpers/Headline';
import { Caption } from 'src/helpers/Caption';
import { BodyCopy } from 'src/helpers/BodyCopy';
import { SvgIcon } from 'src/helpers/SvgIcon';
import classNames from 'classnames';
import 'react-before-after-slider-component/dist/build.css';
import { ComponentProps } from 'lib/types/component-props';
import { useModalIdContext } from 'lib/context/GenericModalIDContext';

type layoutStyle = 'full-width' | 'side-by-side';

export type PromoImageWithSliderProps =
  Feature.EnterpriseWeb.Enterprise.Components.Promo.PromoImageWithSlider.PromoImageWithSlider &
    ComponentProps;

const PromoImageWithSlider = (props: PromoImageWithSliderProps) => {
  const { themeData, themeName } = useTheme(PromoImageWithSliderTheme);
  const { currentScreenWidth } = useCurrentScreenType();

  const [sliderValue, setSliderValue] = useState(50);
  const [showDetails, setShowDetails] = useState(false);
  const [forceUpdateToggle, setForceUpdateToggle] = useState(false);
  const sliderRef = useRef<HTMLDivElement | null>(null);
  const { selectedModalId } = useModalIdContext();

  useEffect(() => {
    if (props.isGenericModal && selectedModalId != '') {
      // Toggle the force update state to trigger a re-render
      setForceUpdateToggle((prevToggle) => !prevToggle);
    }
  }, [props.isGenericModal, selectedModalId]);

  useEffect(() => {
    setForceUpdateToggle((prevToggle) => !prevToggle);
  }, [props.forceUpdate]);

  useEffect(() => {
    // setTimeout because once slider-component gets rerendered, we will have access to image elements in dom
    setTimeout(() => {
      const secondImage = sliderRef.current?.querySelector(
        '.before-after-slider__first-photo-container'
      );

      const firstImage = sliderRef.current?.querySelector(
        '.before-after-slider__second-photo-container'
      );

      secondImage &&
        secondImage?.classList.add(
          'after:block',
          'after:py-s',
          'after:text-center',
          'after:min-w-[160px]',
          'after:bg-black',
          'after:text-white',
          'after:uppercase',
          'after:w-fit',
          'after:content-none',
          `${
            props.fields?.rightImageLabel?.value
              ? 'md:after:content-[attr(data-afterLabel)]'
              : 'md:after:content-[""]'
          }`
        );
      firstImage &&
        firstImage?.classList.add(
          'after:block',
          'after:py-s',
          'after:text-center',
          'after:min-w-[160px]',
          'after:bg-black',
          'after:text-white',
          'after:uppercase',
          'after:w-fit',
          'after:content-none',
          `${
            props.fields?.leftImageLabel?.value
              ? 'md:after:content-[attr(data-beforeLabel)]'
              : 'md:after:content-[""]'
          }`
        );

      secondImage?.setAttribute('data-afterLabel', props.fields?.rightImageLabel?.value || '');
      firstImage?.setAttribute('data-beforeLabel', props.fields?.leftImageLabel?.value || '');
    }, 100);
    // "props.fields.rightImageLabel.value" and "props.fields.leftImageLabel.value" are directly coming from layout service.
    // We can ignore the react-hooks/exhaustive-deps warning for this useEffect.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [sliderRef.current, forceUpdateToggle]);

  const LAYOUT_STYLE = getEnum<layoutStyle>(props.fields?.layoutStyle) || 'full-width';

  const FIRST_IMAGE = {
    imageUrl: props.fields?.rightImage.value?.src || '',
    alt: props.fields?.rightImage.value?.alt as string,
  };

  const SECOND_IMAGE = {
    imageUrl: props.fields?.leftImage.value?.src || '',
    alt: props.fields?.leftImage.value?.alt as string,
  };

  const delimiterIconStyles = {
    width: '48px',
    height: '48px',
    display: currentScreenWidth < getBreakpoint('md') ? 'none' : 'block',
    backgroundPosition: 'center',
    borderRadius: 'none',
    backgroundColor: themeName === 'aw' ? '#F26924' : '#000000',
    backgroundRepeat: 'no-repeat',
    // Note: using icons file from public folder directly as we can only use background-image property
    // to customise slider thumb (delimiterIconStyles)
    backgroundImage:
      themeName === 'aw'
        ? "url('/before-after-slider-arrow-aw.svg')"
        : "url('/before-after-slider-arrow-rba.svg')",
  };

  let promoContentColumn;
  let promoImageColumn;

  switch (LAYOUT_STYLE) {
    case 'side-by-side':
      promoContentColumn = themeData.classes.sideBySide.promoContent;
      promoImageColumn = themeData.classes.sideBySide.promoImgSlider;
      break;

    case 'full-width':
      promoContentColumn = themeData.classes.fullWidth.promoContent;
      promoImageColumn = themeData.classes.fullWidth.promoImgSlider;
      break;
  }

  function handleKeyDown(e: KeyboardEvent) {
    if (e.key === 'ArrowLeft' && sliderValue > -1) {
      setSliderValue(sliderValue - 1);
    }
    if (e.key === 'ArrowRight' && sliderValue < 100) {
      setSliderValue(sliderValue + 1);
    }
  }

  const renderSlider = (): JSX.Element => {
    return (
      <div>
        {LAYOUT_STYLE === 'full-width' && (
          <Headline
            {...props}
            classes={classNames(themeData.classes.fullWidth.headline, 'mb-xxs md:mb-s')}
          />
        )}

        <div
          tabIndex={currentScreenWidth < getBreakpoint('md') ? -1 : 0}
          onKeyDown={handleKeyDown}
          className="relative -mx-m md:mx-0"
        >
          {/* If rendered within generic modal, force rerender on modal open as it does not get initialised on first load */}
          <ReactBeforeSliderComponent
            key={
              props.isGenericModal || props.forceUpdate
                ? `forceRerender-${forceUpdateToggle}`
                : 'react-before-after-slider'
            }
            currentPercentPosition={sliderValue}
            onChangePercentPosition={(position) => {
              // returning false to prevent thumb slide in mobile
              if (currentScreenWidth < getBreakpoint('md')) {
                return false;
              } else {
                setSliderValue(position);
                return true;
              }
            }}
            firstImage={FIRST_IMAGE}
            secondImage={SECOND_IMAGE}
            feelsOnlyTheDelimiter={true}
            delimiterIconStyles={delimiterIconStyles}
            delimiterColor="white"
          />
          <div className={themeData.classes.rangewrapper}>
            <Text
              field={props.fields?.rightImageLabel}
              tag="span"
              className={classNames(themeData.classes.imageLabels, 'left-[30px]')}
            />
            {currentScreenWidth < getBreakpoint('md') && (
              <>
                <input
                  aria-label="before-after range input"
                  className="before-after-slider-range mx-xxs"
                  onChange={(e) => setSliderValue(parseFloat(e.target.value))}
                  type="range"
                  min={-1}
                  max={100}
                  value={sliderValue}
                />
              </>
            )}
            <Text
              field={props.fields?.leftImageLabel}
              tag="span"
              className={classNames(themeData.classes.imageLabels, 'right-[30px]')}
            />
          </div>
        </div>
      </div>
    );
  };

  const expandDetailsButton = () => {
    const COLLAPSED_LABEL =
      themeName === 'aw' ? { value: 'SHOW DETAILS' } : { value: 'View Details' };
    const EXPANDED_LABEL =
      themeName === 'aw' ? { value: 'CLOSE DETAILS' } : { value: 'Hide Details' };
    return (
      <button className={themeData.classes.expandButton.wrapper}>
        <SvgIcon
          icon={showDetails ? 'close' : 'plus'}
          size={themeName === 'aw' ? 'sm' : 'md'}
          className={themeData.classes.expandButton.icon}
        />
        <Text
          tag="span"
          field={showDetails ? EXPANDED_LABEL : COLLAPSED_LABEL}
          className={themeData.classes.expandButton.text}
        />
      </button>
    );
  };

  const renderAdditionalDetails = () => {
    if (LAYOUT_STYLE === 'full-width') {
      return (
        <div className="grid grid-cols-12 ">
          <div className="col-span-12 md:col-span-10">
            <Caption
              caption={props.fields?.imageCaption}
              italic={false}
              isImageCaption={false}
              classes={themeData.classes.caption}
            />
          </div>

          <div
            className="col-span-12 md:col-span-2 md:justify-self-end "
            onClick={() => setShowDetails(!showDetails)}
          >
            {expandDetailsButton()}
          </div>

          {showDetails && (
            <BodyCopy
              fields={{ body: props.fields?.description || { value: '' } }}
              classes={themeData.classes.bodyCopy}
            />
          )}
        </div>
      );
    } else {
      return (
        <>
          <div className="order-last md:order-none">
            <Headline
              {...props}
              classes={classNames(themeData.classes.sideBySide.headline, 'mb-xxs md:mb-s ')}
            />
            <BodyCopy
              fields={{ body: props.fields?.description || { value: '' } }}
              classes={themeData.classes.bodyCopy}
            />
          </div>

          <Caption
            caption={props.fields?.imageCaption}
            italic={false}
            classes={classNames(themeData.classes.caption, themeName === 'rba' && 'md:mb-0')}
          />
        </>
      );
    }
  };

  return (
    <Component variant="lg" dataComponent="promo/promoimagewithslider" {...props}>
      <div className={promoContentColumn}>{renderAdditionalDetails()}</div>
      <div ref={sliderRef} className={promoImageColumn}>
        {renderSlider()}
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<PromoImageWithSliderProps>(PromoImageWithSlider);
