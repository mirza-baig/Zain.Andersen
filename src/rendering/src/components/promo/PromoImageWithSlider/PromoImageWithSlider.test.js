// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import PromoImageWithSlider from './PromoImageWithSlider';
import defaultData from './PromoImageWithSlider.full-width.mock-data';

it('renders correctly', async () => {
  const component = snapshot(PromoImageWithSlider, { componentProps: defaultData });
  await hasDataComponent(component, 'promo/promoimagewithslider');
});
