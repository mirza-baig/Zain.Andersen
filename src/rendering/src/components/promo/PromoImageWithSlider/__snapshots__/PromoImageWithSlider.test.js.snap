// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="promo/promoimagewithslider"
  id="id151852150"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s "
    >
      <div
        class="col-span-12 order-last"
      >
        <div
          class="grid grid-cols-12 "
        >
          <div
            class="col-span-12 md:col-span-10"
          >
            <div
              class="caption"
            >
              <div
                class="md:mb-0  text-dark-gray mt-xxs text-left mb-s font-regular [&_strong]:text-black  font-serif text-sm-xxs text-body"
              >
                Image caption goes here.
              </div>
            </div>
          </div>
          <div
            class="col-span-12 md:col-span-2 md:justify-self-end "
          >
            <button
              class="flex flex-row md:flex-row-reverse items-center gap-x-xxs  group"
            >
              <span
                class="flex h-m w-m items-center justify-center rounded-full border-2 border-primary hover:bg-primary"
              />
              <span
                class="text-text-link font-sans font-regular group-hover:underline underline-offset-4"
              >
                SHOW DETAILS
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="col-span-12"
      >
        <div>
          <div
            class="items-top font-sans [&_a:hover]:underline text-theme-text text-sm-m md:text-m  font-heavy mb-xxs md:mb-s"
          >
            <h2>
              Headline Text
            </h2>
          </div>
          <div
            class="relative -mx-m md:mx-0"
            tabindex="0"
          >
            <div
              class="before-after-slider"
            >
              <div
                class="before-after-slider__first-photo-container"
              >
                <img
                  alt="woman in entrance of home with andersen black framed door open"
                  draggable="false"
                  src="https://source.unsplash.com/random/1200x675?after"
                />
              </div>
            </div>
            <div
              class="flex items-center justify-center bg-light-gray px-m"
            >
              <span
                class="top-[30px] z-10 inline-block py-s text-center text-sm-xxs md:text-xs font-sans font-heavy text-black md:absolute md:w-[160px] md:bg-black md:text-white uppercase md:hidden left-[30px]"
              >
                After
              </span>
              <span
                class="top-[30px] z-10 inline-block py-s text-center text-sm-xxs md:text-xs font-sans font-heavy text-black md:absolute md:w-[160px] md:bg-black md:text-white uppercase md:hidden right-[30px]"
              >
                Before
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
